moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+      method.isAccessible = true
+      val returnType = method.genericReturnType
+      val parameterTypes = method.genericParameterTypes
+      val parameterAnnotations = method.parameterAnnotations
+      return if (parameterTypes.size >= 2 &&
+        parameterTypes[0] == JsonWriter::class.java &&
+        returnType == Void.TYPE &&
+        parametersAreJsonAdapters(2, parameterTypes)
+      ) {
+        // void pointToJson(JsonWriter jsonWriter, Point point) {
+        // void pointToJson(JsonWriter jsonWriter, Point point, JsonAdapter<?> adapter, ...) {
+        val qualifierAnnotations = parameterAnnotations[1].jsonAnnotations
+        object : AdapterMethod(
+          parameterTypes[1],
+          qualifierAnnotations,
+          adapter,
+          method,
+          parameterTypes.size,
+          2,
+          true
+        ) {
+          override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+            invoke(writer, value)
+          }
+        }
+      } else if (parameterTypes.size == 1 && returnType != Void.TYPE) {
+        // List<Integer> pointToJson(Point point) {
+        val returnTypeAnnotations = method.jsonAnnotations
+        val qualifierAnnotations = parameterAnnotations[0].jsonAnnotations
+        val nullable = parameterAnnotations[0].hasNullable
+        object : AdapterMethod(
+          parameterTypes[0],
+          qualifierAnnotations,
+          adapter,
+          method,
+          parameterTypes.size,
+          1,
+          nullable
+        ) {
+          private var delegate: JsonAdapter<Any>? = null
+          override fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+            super.bind(moshi, factory)
+            delegate = if (Types.equals(parameterTypes[0], returnType) &&
+              qualifierAnnotations == returnTypeAnnotations
+            )
+              moshi.nextAdapter(factory, returnType, returnTypeAnnotations)
+            else moshi.adapter(returnType, returnTypeAnnotations)
+          }
+
+          override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+            val intermediate = invoke(value)
+            delegate!!.toJson(writer, intermediate)
+          }
+        }
+      } else {
+        throw IllegalArgumentException(
+          """Unexpected signature for $method.
+@ToJson method signatures may have one of the following structures:
+    <any access modifier> void toJson(JsonWriter writer, T value) throws <any>;
+    <any access modifier> void toJson(JsonWriter writer, T value, JsonAdapter<any> delegate, <any more delegates>) throws <any>;
+    <any access modifier> R toJson(T value) throws <any>;
+"""
+        )
+      }
+    }
+
+    /** Returns true if `parameterTypes[offset]` contains only JsonAdapters.  */
+    private fun parametersAreJsonAdapters(offset: Int, parameterTypes: Array<Type>): Boolean {
+      var i = offset
+      val length = parameterTypes.size
+      while (i < length) {
```

Comment:
Can this be a for-loop?

Reply:
https://github.com/square/moshi/pull/1467/commits/3e6e9bda3372997dff18ac5d1567f6b80707074e#diff-e8a15c8ca843b45b53cf5d1b2097710c70dd4b5a9399c074e7fa2b379a772d47R215